### PR TITLE
Bug 1712731 - Compute the noise profile of the data when generating alerts.

### PR DESCRIPTION
**Pull request by gmierz**
This patch integrates the measure-noise tool into perfherder. It makes use of the deviance method to determine the noise profile of the data in a given alert. With this patch, the alerts will be annotated with a `noise_profile` (described [here](https://github.com/mozilla/measure-noise)) that will be used in the frontend in the future in the alerts view.

The data used to compute this profile includes everything after the data series' previous alert and before the new alert being created. Everything before the previous alert possibly follows a different profile, and everything after the new alert may also have a shift in the noise profile.

### CODE REVIEW

File: treeherder/perf/models.py [hunk 1]
```diff
@@ -447,6 +447,11 @@ class PerformanceAlert(models.Model):
         help_text="t value out of analysis indicating confidence " "that change is 'real'",
         null=True,
     )
+    noise_profile = models.CharField(
```
Comment (beatrice-acasandrei) [inline treeherder/perf/models.py:450]: I think it could be useful to define the [choices](https://docs.djangoproject.com/en/3.2/ref/models/fields/#choices) for the field `noise_profile` :

- `SKEWED` - samples are heavily to one side of the mean
- `OUTLIERS` - there are more outliers than would be expected from normal distribution
- `MODAL` - few samples are near the mean (probably bimodal)
- `OK` - no egregious deviation from normal
- `N/A` - not enough data to make a conclusion (aka OK)

File: treeherder/perf/alerts.py [hunk 2]
```diff
@@ -83,6 +96,37 @@ def generate_new_alerts_in_series(signature):
...
+                    # shows the regression and obtain a noise profile on it.
+                    # This helps us to ignore this alert and others in the
+                    # calculation that could influence the profile.
+                    noise_data = []
+                    for point in analyzed_series:
+                        if point == cur:
+                            break
+                        noise_data.append(geomean(point.values))
+
+                    noise_profile = deviance(noise_data)
```
Comment (beatrice-acasandrei) [inline treeherder/perf/alerts.py:111]: I've noticed that `noise_profile[1]` is not used. Instead of `noise_profile = deviance(noise_data)` you could do `noise_profile, _ = deviance(noise_data)` and then use `noise_profile` instead of `noise_profile[0]` where needed. Hope it makes sense.

**Review by beatrice-acasandrei — APPROVED** [review]
Looks good to me!

**Review by airimovici — APPROVED** [review]
It only needs the linter fixes (`flake8` and `black`) and it should be good

**Comment by beatrice-acasandrei** [comment]
Hello @sarah-clements! This PR adds a new field to the `PerformanceAlert` model, could you please take a look on the changes? Thanks.

File: requirements/common.txt [hunk 3]
```diff
@@ -127,7 +127,7 @@ cachy==0.3.0 \
 celery==4.4.2 \
     --hash=sha256:108a0bf9018a871620936c33a3ee9f6336a89f8ef0a0f567a9001f4aa361415f \
     --hash=sha256:5b4b37e276033fe47575107a2775469f0b721646a08c96ec2c61531e4fe45f2a
-    # via -r common.in
+    # via -r requirements/common.in
```
Comment (sarah-clements) [inline requirements/common.txt:130]: Why was this change included in this pr?

File: treeherder/perf/alerts.py [hunk 4]
```diff
@@ -83,6 +96,29 @@ def generate_new_alerts_in_series(signature):
...
+                    noise_profile, _ = deviance(noise_data)
+
+                    if not isinstance(noise_profile, str):
+                        raise Exception(
+                            "Expecting a string as a " f"noise profile, got: {type(noise_profile)}"
+                        )
+                except Exception:
+                    # Fail without breaking the alert computation
+                    logger.warning("Failed to obtain a noise profile.")
+                    traceback.print_exc()
```
Comment (sarah-clements) [inline treeherder/perf/alerts.py:120]: You could add new relic reporting here too or instead. There should be examples in the codebase.

File: treeherder/perf/migrations/0040_performancealert_noise_profile.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+                    ),
+                    ('OK', 'No issues were found.'),
+                    ('N/A', 'Could not compute a noise profile.'),
+                ],
+                default='N/A',
+                help_text='The noise profile of the data which precedes this alert.',
+                max_length=30,
+            ),
+        ),
+    ]
```
Comment (sarah-clements) [inline treeherder/perf/migrations/0040_performancealert_noise_profile.py:35]: Has the django migration/schema change been tested on prototype? If its a very large table, it might be a good idea to do so before a prod deploy.

**Comment by sarah-clements** [comment]
This looks good but when you're ready to merge you might want to coordinate with @bhearsum. He's in charge of prod deploys while I'm on PTO (7/22 through Aug 1st).

**Review by sarah-clements — APPROVED** [review]

File: requirements/common.txt (hunk 3 above)
Comment (beatrice-acasandrei) [reply requirements/common.txt:130]: We're not sure why this change was included and Greg has PTO.
